api/envoy/extensions/clusters/dynamic_forward_proxy/v3/cluster.proto

```diff
@@ -32,4 +32,16 @@ message ClusterConfig {
   // in the :ref:`cluster's upstream_http_protocol_options
   // <envoy_v3_api_field_config.cluster.v3.Cluster.upstream_http_protocol_options>`
   bool allow_insecure_cluster_options = 2;
+
+  // If true allow HTTP/2 and HTTP/3 connections to be reused for requests to different
+  // origins than the connection was initially created for. This will only happen when the
+  // resolved address for the new connection matches the peer address of the connection and
+  // the TLS certificate is also valid for the new hostname. For example, if a connection
+  // has previously been established to foo.example.com at IP 1.2.3.4 with a certificate
+  // that is valid for `*.example.com`, then this connection could be used for requests to
+  // bar.example.com if that also resolved to 1.2.3.4.
+  //
+  // NOTE: The coalesced connections might be to upstreams that would not be otherwise
```

Comment:
This seems ominent.. what does the note imply? It's a bit too broad a claim.

Also, please use RST syntax for notes (see https://sublime-and-sphinx-guide.readthedocs.io/en/latest/notes_warnings.html).

Reply:
Thanks for the RST note syntax. Done!

As for the claim, I was attempt to answer the concern here:

https://github.com/envoyproxy/envoy/pull/17874#discussion_r698856052

Do you have a suggestion on how to phrase it better? The note is trying to say that if connection coalescing results in a request to "foo" being sent on a connection to "bar", that would be different than the host otherwise selected by envoy. Maybe I need to be more specific?

Reply:
Yeah, I'd just say something like "this might violate RFC 7540, see https://github.com/envoyproxy/envoy/issues/6767" or something like that

Reply:
Hm. I think #6767 is complaining about Envoy not sending a 421 downstream when a client reuses a connection to Envoy for different origin that Envoy does not expect on that connection.

This PR changes the upstream connection logic. I understood the concern in the comment to be that even though we have an upstream connection with a valid cert for `*.example.com`, Envoy would not normally send a request for a different origin down that connection and this may be surprising. So I don't think enabling coalescing will cause RFC 7540 violations so much as it override some possible Envoy upstream configuration.

At least, that's what I think. Does that make sense?

Reply:
Yeah, you're right, this isn't exactly #6767. It's also probably not a technical RFC 7540, but do you think https://datatracker.ietf.org/doc/html/rfc7540#section-9.1.1 describes the underlying issue here?

I.e. we're reusing a connectino and might hit a non-authoritative server by accident. Would linking to that make sense?

Reply:
Sure! I've added a link, PTAL.